src/CSEjector.sol

```diff
@@ -0,0 +1,128 @@
+// SPDX-FileCopyrightText: 2025 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.24;
+
+import { PausableUntil } from "./lib/utils/PausableUntil.sol";
+import { AccessControlEnumerableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/extensions/AccessControlEnumerableUpgradeable.sol";
+import { Initializable } from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+import { IStakingModule } from "./interfaces/IStakingModule.sol";
+import { ICSParametersRegistry } from "./interfaces/ICSParametersRegistry.sol";
+import { ICSAccounting } from "./interfaces/ICSAccounting.sol";
+import { ICSModule, NodeOperator } from "./interfaces/ICSModule.sol";
+import { ICSEjector } from "./interfaces/ICSEjector.sol";
+
+import { AssetRecoverer } from "./abstract/AssetRecoverer.sol";
+
+contract CSEjector is
+    ICSEjector,
+    Initializable,
+    AccessControlEnumerableUpgradeable,
+    PausableUntil,
+    AssetRecoverer
+{
+    bytes32 public constant PAUSE_ROLE = keccak256("PAUSE_ROLE");
+    bytes32 public constant RESUME_ROLE = keccak256("RESUME_ROLE");
+    bytes32 public constant BAD_PERFORMER_EJECTOR_ROLE =
+        keccak256("BAD_PERFORMER_EJECTOR_ROLE");
+    bytes32 public constant RECOVERER_ROLE = keccak256("RECOVERER_ROLE");
+
+    ICSModule public immutable MODULE;
+    ICSAccounting public immutable ACCOUNTING;
+
+    /// @dev see _keyPointer function for details of noKeyIndexPacked structure
+    mapping(uint256 noKeyIndexPacked => bool) private _isValidatorEjected;
+
+    constructor(address module) {
+        if (module == address(0)) revert ZeroModuleAddress();
+        MODULE = ICSModule(module);
+        ACCOUNTING = ICSAccounting(MODULE.accounting());
+    }
+
+    /// @notice initialize the contract from scratch
+    function initialize(address admin) external initializer {
+        if (admin == address(0)) revert ZeroAdminAddress();
+
+        __AccessControlEnumerable_init();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+    }
+
+    /// @inheritdoc ICSEjector
+    function resume() external onlyRole(RESUME_ROLE) {
+        _resume();
+    }
+
+    /// @inheritdoc ICSEjector
+    function pauseFor(uint256 duration) external onlyRole(PAUSE_ROLE) {
+        _pauseFor(duration);
+    }
+
+    /// @inheritdoc ICSEjector
+    function ejectBadPerformer(
+        uint256 nodeOperatorId,
+        uint256 keyIndex,
+        uint256 strikesCount
+    ) external whenResumed onlyRole(BAD_PERFORMER_EJECTOR_ROLE) {
+        _onlyExistingNodeOperator(nodeOperatorId);
+        NodeOperator memory no = MODULE.getNodeOperator(nodeOperatorId);
```

Comment:
`getNodeOperatorSummary` should be cheaper

Reply:
or consider even adding a new view method

Reply:
A separate method added